File: aes_ctr.py
from Cryptodome.Cipher import AES
from Cryptodome import Random
from Cryptodome.Util import Counter
import hashlib
from base64 import b64decode, b64encode



class AES_CTR:

    def __init__(self, key):
        self.key = hashlib.sha256(key.encode()).digest()

    def encrypt(self,plain_text):
        # Generate random iv eg: nonce as a series of bytes 
        iv = Random.get_random_bytes(12)
        # creating a counter block  : first arg is the length of counter value to be
        # concated with nonce(prefix). The length of both combined should be equal to 128 bits
        counter = Counter.new(32, prefix= iv)
        # create a cipher with ctr mode using the cointer object
        cipher = AES.new(self.key, AES.MODE_CTR, counter= counter)
        # encrypt
        encrypted = cipher.encrypt(plain_text.encode())
        
        return b64encode(iv + encrypted).decode() 

    def decrypt(self, encrypted_text):

        # decode from the b64encoding
        encrypted_text  = b64decode(encrypted_text.encode())
        # generate prepended iv from the e-text and cipher text
        iv  = encrypted_text[:12]
        cipher_text = encrypted_text[12:]
        # create counter object
        counter = Counter.new(32, prefix = iv )
        cipher = AES.new(self.key, AES.MODE_CTR, counter = counter)
        plain_text = cipher.decrypt(cipher_text)
        return plain_text.decode()  


      

        

File: aesCTR.js
const crypto = require('node:crypto')

class AES_CTR {

    constructor(key) {
        this.key = crypto.createHash('sha256').update(key).digest()
        this.algorithm = 'aes-256-ctr'
    }

    encrypt = (plainText) => {
        // generate initialization vector
        const iv = crypto.randomBytes(16)
        // create a cipher using the  aes-256-ctr
        const cipher = crypto.createCipheriv(this.algorithm, this.key, iv)
        // encrypt the text with input as utf-8 to output base64
        let encryptedText = cipher.update(plainText, 'utf-8', 'base64')
        // finalize the reamining part of the encrypted text
        encryptedText += cipher.final('base64')
        // concaatenate the iv to the cipher text
        const encodedBytes = Buffer.concat([iv, Buffer.from(encryptedText, 'base64')])
        // encode the bytes to a more transmittable format which is the base64 format 
        return encodedBytes.toString('base64')
    }
}

aes = new AES_CTR('1234')

console.log(aes.encrypt('wadii wangu'))